File: dispatch/java/runtime/adamma/c4dhi/claid/Module/Subscriber.java
/***************************************************************************
* Copyright (C) 2023 ETH Zurich
* CLAID: Closing the Loop on AI & Data Collection (https://claid.ethz.ch)
* Core AI & Digital Biomarker, Acoustic and Inflammatory Biomarkers (ADAMMA)
* Centre for Digital Health Interventions (c4dhi.org)
* 
* Authors: Patrick Langer
* 
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
* 
*         http://www.apache.org/licenses/LICENSE-2.0
* 
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
***************************************************************************/

package adamma.c4dhi.claid.Module;

import java.time.Instant;
import java.time.LocalDateTime;
import java.time.ZoneId;
import java.util.function.Consumer;

import adamma.c4dhi.claid.Logger.Logger;
import adamma.c4dhi.claid.Module.Scheduling.ConsumerRunnable;
import adamma.c4dhi.claid.Module.Scheduling.RunnableDispatcher;
import adamma.c4dhi.claid.Module.Scheduling.ScheduleOnce;
import adamma.c4dhi.claid.Module.Scheduling.ScheduledRunnable;
import adamma.c4dhi.claid.TypeMapping.DataType;
import adamma.c4dhi.claid.TypeMapping.Mutator;
import adamma.c4dhi.claid.TypeMapping.TypeMapping;
import adamma.c4dhi.claid.DataPackage;


public class Subscriber<T extends Object> extends AbstractSubscriber
{
    private DataType dataType;
    private Consumer<ChannelData<T>> callback;
    private RunnableDispatcher callbackDispatcher;

    private Mutator<T> mutator;
        
    public Subscriber(DataType dataType, Consumer<ChannelData<T>> callback, RunnableDispatcher callbackDispatcher)
    {
        this.dataType = dataType;
        this.callback = callback;
        this.callbackDispatcher = callbackDispatcher;
        this.mutator = TypeMapping.getMutator(dataType);
    }

    private void invokeCallback(ChannelData<T> data)
    {
        ConsumerRunnable<ChannelData<T>> consumerRunnable = new ConsumerRunnable<ChannelData<T>>(data, callback, ScheduleOnce.now());
        ScheduledRunnable runnable = (ScheduledRunnable) consumerRunnable;

        this.callbackDispatcher.addRunnable(runnable);
    }

    @Override
    public void onNewData(DataPackage data) 
    {
        T value = this.mutator.getPackagePayload(data);

        Instant instant = Instant.ofEpochSecond(data.getUnixTimestampMs());

        // Convert Instant to LocalDateTime
        LocalDateTime dateTime = LocalDateTime.ofInstant(instant, ZoneId.systemDefault());

        ChannelData channelData = new ChannelData(value, dateTime, data.getSourceUserToken());

        this.invokeCallback(channelData);

        /*if(dataType.isInstance(data))
        {   
            T typedData = dataType.cast(data);
            invokeCallback(typedData);
        }
        else
        {
            Logger.logError("Error in Java Runtime: Subscriber received data of type \"" + data.getClass().getName() + "\", but\n"
            + "expected data of type \"" + dataType.getName() + "\". Cast failed.");
        }*/
    }

}

File: dispatch/python/claid/module/subscriber.py
###########################################################################
# Copyright (C) 2023 ETH Zurich
# CLAID: Closing the Loop on AI & Data Collection (https://claid.ethz.ch)
# Core AI & Digital Biomarker, Acoustic and Inflammatory Biomarkers (ADAMMA)
# Centre for Digital Health Interventions (c4dhi.org)
# 
# Authors: Patrick Langer
# 
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
# 
#         http://www.apache.org/licenses/LICENSE-2.0
# 
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
##########################################################################

from module.abstract_subscriber import AbstractSubscriber
from module.scheduling.runnable_dispatcher import RunnableDispatcher

from module.scheduling.schedule_once import ScheduleOnce
from module.channel_data import ChannelData
from module.type_mapping.mutator import Mutator
from module.type_mapping.type_mapping import TypeMapping
from module.scheduling.function_runnable_with_params import FunctionRunnableWithParams
from module.scheduling.scheduled_runnable import ScheduledRunnable
from datetime import datetime
class Subscriber(AbstractSubscriber):
    def __init__(self, data_type_example_instance, callback, callback_dispatcher):
        self.callback = callback
        self.callback_dispatcher = callback_dispatcher
        self.mutator = TypeMapping.get_mutator(data_type_example_instance)

    def invoke_callback(self, data):
        function_runnable = FunctionRunnableWithParams(self.callback)
        function_runnable.set_params(data)

        
        self.callback_dispatcher.add_runnable(
            ScheduledRunnable(
                function_runnable,
                ScheduleOnce(datetime.now())
            )
        )

    def on_new_data(self, package):

        data = self.mutator.get_package_payload(package)
        channel_data = ChannelData( 
            data,
            datetime.fromtimestamp(package.unix_timestamp_ms / 1000.0),
            package.source_user_token
        )

        self.invoke_callback(channel_data)
